src/main/scala/akka/stream/contrib/PartitionWith.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.japi.function
+import akka.stream.{ Attributes, FanOutShape2 }
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+
+/**
+ * This companion defines a factory for [[PartitionWith]] instances, see [[PartitionWith.apply]].
+ */
+object PartitionWith {
+
+  /**
+   * Factory for [[PartitionWith]] instances.
+   *
+   * @param p partition function
+   * @tparam In input type
+   * @tparam Out0 left output type
+   * @tparam Out1 right output type
+   * @return [[PartitionWith]] instance
+   */
+  def apply[In, Out0, Out1](p: In => Either[Out0, Out1]): PartitionWith[In, Out0, Out1] = new PartitionWith(p)
+
+  /**
+   * Java API: Factory for [[PartitionWith]] instances.
+   *
+   * @param p partition function
+   * @tparam In input type
+   * @tparam Out0 left output type
+   * @tparam Out1 right output type
+   * @return [[PartitionWith]] instance
+   */
+  def create[In, Out0, Out1](p: function.Function[In, Either[Out0, Out1]]): PartitionWith[In, Out0, Out1] = new PartitionWith(p.apply)
+}
+
+/**
+ * This stage partitions input to 2 different outlets,
+ * applying different transformations on the elements,
+ * according to the received partition function.
+ *
+ * @param p partition function
+ * @tparam In input type
+ * @tparam Out0 left output type
+ * @tparam Out1 right output type
+ */
+final class PartitionWith[In, Out0, Out1] private (p: In => Either[Out0, Out1]) extends GraphStage[FanOutShape2[In, Out0, Out1]] {
+
+  override val shape = new FanOutShape2[In, Out0, Out1]("partitionWith")
+
+  override def createLogic(attributes: Attributes) = new GraphStageLogic(shape) {
+    import shape._
+
+    private var pending = null.asInstanceOf[Either[Out0, Out1]]
```

Comment:
it would be more clean with 
```
private var pending: [Either[Out0, Out1]] = null
```

then you can also use `pending = null` below at two places